File: js/day01/day1.js
import fs from 'fs'
import path from 'path'

let filename = path.join(path.resolve(), 'data')

let data = new Set(
    fs.readFileSync(filename, 'utf8')
        .split('\n')
        .filter(x => x !== '')
        .map(item => parseInt(item.trim()))
)

function find_group(nums, n, count) {
    if (n < 0) {
        return null
    }
    for (let i of nums) {
        if (count <= 2) {
            if (nums.has(n-i)) {
                return [i, n-i]
            }
        }
        else {
            let copy = new Set(nums)
            copy.delete(i)
            let possible = find_group(copy, n-i, count-1)
            if (possible) {
                possible.push(i)
                return possible
            }
        }
    }
    return null
}

console.log(find_group(data, 2020, 2))
console.log(find_group(data, 2020, 3))

File: python/day01/day1.py
import pathlib
from functools import reduce

CWD = pathlib.Path(__file__).parent.absolute()
filename = pathlib.PurePath(CWD, 'data')
with open(filename) as f:
    nums = set(int(x.strip()) for x in f.readlines())

def find_group(nums, n, count):
    if n < 0:
        return None
    for i in nums:
        nums.remove(i)
        if count <= 2:
            if n-i in nums:
                return [i, n-i]
        else:
            possible = find_group(nums, n-i, count-1)
            if possible:
                return [i] + possible
        nums.add(i)


group = find_group(nums, 2020, 2)
print(f'Part 1: {reduce(lambda x, y: x*y, group)}')
group = find_group(nums, 2020, 3)
print(f'Part 2: {reduce(lambda x, y: x*y, group)}')
